File: Poisson_equation_2D.m

function [n2]=Poisson_equation_2D(Lx,Lz,Rhs,Const)

% Solving the 2-D Poisson equation by the Finite Difference
...Method 
% Numerical scheme used is a second order central difference in space
...(5-point difference)
   

[Nx,Nz]=size(Rhs);

%
%Specifying parameters (check why It does not work if we change them)

dx=Lx/(Nx-1);                      %Width of space step(x)
dy=Lz/(Nz-1);                      %Width of space step(y)
x=0:dx:Lx;                        %Range of x(0,2) and specifying the grid points
y=0:dy:Lz;                        %Range of y(0,2) and specifying the grid points
b=zeros(Nx,Nz);                    %Preallocating b
pn=zeros(Nx,Nz);                   %Preallocating pn

%%
% Initial Conditions
p=zeros(Nx,Nz);                  %Preallocating p

%% 
Rhs=Const.*fliplr(Rhs);   
b=(Rhs);

%%
i=2:Nx-1;
j=2:Nz-1;

% Poisson equation solution (iterative) method

tol = 1e-4;			 % Set tollerance
maxerr = inf;	     % initial error
iter = 0;
pn=p;

while maxerr > tol
    iter = iter + 1;
    %disp(['Iteration no. ',num2str(iter)]);

    
    %Explicit iterative scheme with C.D in space (5-point difference)

    p(i,j)=((dy^2*(pn(i+1,j)+pn(i-1,j)))+(dx^2*(pn(i,j+1)+pn(i,j-1)))-(b(i,j)*dx^2*dy*2))/(2*(dx^2+dy^2));
    
%Boundary conditions     
%% Neumann's conditions   % dp/dx|end=dp/dx|end-1 
    p(1,:)=p(2,:);
    p(end,:)=p(end-1,:);
    
    
%%    Dirichlet's conditions 
%     p(:,1)=1.43;
%     p(:,end)=1.32;
    
%% Neumann's conditions
    p(:,1)=p(:,2);
    p(:,end)=p(:,end-1);

    maxerr = max(max(abs((p-pn)./p)));
    %disp(['Maximum error is  ',num2str(maxerr)]);
    pn=p;
   
end			% as long the error larger than tolerance, continue


PG2_gray=p*255;
n_max=1.43;    
n_min=1.332;
n2= scaledata(PG2_gray,n_min,n_max);
 


% Check the shape fucntion

% figure 
% subplot(121)
% h=surf(x,y,p','EdgeColor','none');       
% shading interp
% %axis([-0.5 2.5 -0.5 2.5 -100 100])
% title({'2-D Poisson equation';['{\itNumber of iterations} = ',num2str(iter)]})
% xlabel('Spatial co-ordinate (x) \rightarrow')
% ylabel('{\leftarrow} Spatial co-ordinate (y)')
% zlabel('Solution profile (P) \rightarrow')
% 
% subplot(122)
% h=surf(x,y,n2','EdgeColor','none');       
% shading interp
% %axis([-0.5 2.5 -0.5 2.5 -100 100])
% title({'2-D Poisson equation';['{\itNumber of iterations} = ',num2str(iter)]})
% xlabel('Spatial co-ordinate (x) \rightarrow')
% ylabel('{\leftarrow} Spatial co-ordinate (y)')
% zlabel('Solution profile (P) \rightarrow')
% 
% 
% % Plot a scaled profile 
% figure
% plot(n2(5,:),y)

File: poisson_equation_2d.py
import numpy as np
from scaledata import scaledata

def Poisson_equation_2D(Lx, Lz, Rhs, Const):
    """
    Solve the 2D Poisson equation using the Finite Difference Method
    
    Parameters:
    -----------
    Lx : int
        Width of the domain in pixels
    Lz : int
        Height of the domain in pixels
    Rhs : ndarray
        Right-hand side of the Poisson equation
    Const : float
        Constant for the Poisson equation
        
    Returns:
    --------
    n2 : ndarray
        Solution of the Poisson equation (refractive index field)
        
    Notes:
    ------
    Numerical scheme used is a second order central difference in space
    (5-point difference)
    """
    # Get dimensions of RHS
    Nx, Nz = Rhs.shape
    
    # Specify parameters
    dx = Lx / (Nx - 1)  # Width of space step (x)
    dy = Lz / (Nz - 1)  # Width of space step (y)
    x = np.linspace(0, Lx, Nx)  # Range of x and grid points
    y = np.linspace(0, Lz, Nz)  # Range of y and grid points
    
    # Preallocate arrays
    b = np.zeros((Nx, Nz))
    pn = np.zeros((Nx, Nz))
    
    # Initial conditions
    p = np.zeros((Nx, Nz))
    
    # Prepare RHS
    Rhs = Const * np.fliplr(Rhs)
    b = Rhs
    
    # Define interior points
    i = np.arange(1, Nx-1)
    j = np.arange(1, Nz-1)
    
    # Poisson equation solution (iterative method)
    tol = 1e-4  # Set tolerance
    maxerr = float('inf')  # Initial error
    iter_count = 0
    pn = p.copy()
    
    # Iterative solution
    while maxerr > tol:
        iter_count += 1
        
        # Explicit iterative scheme with central difference in space (5-point difference)
        for ii in i:
            for jj in j:
                p[ii, jj] = ((dy**2 * (pn[ii+1, jj] + pn[ii-1, jj])) + 
                             (dx**2 * (pn[ii, jj+1] + pn[ii, jj-1])) - 
                             (b[ii, jj] * dx**2 * dy * 2)) / (2 * (dx**2 + dy**2))
        
        # Boundary conditions
        # Neumann's conditions: dp/dx|end = dp/dx|end-1
        p[0, :] = p[1, :]
        p[-1, :] = p[-2, :]
        
        # Neumann's conditions
        p[:, 0] = p[:, 1]
        p[:, -1] = p[:, -2]
        
        # Calculate error
        with np.errstate(divide='ignore', invalid='ignore'):
            err_matrix = np.abs((p - pn) / p)
            err_matrix[~np.isfinite(err_matrix)] = 0
            maxerr = np.max(err_matrix)
        
        # Update previous solution
        pn = p.copy()
    
    # Scale the solution to physical values
    PG2_gray = p * 255
    n_max = 1.43
    n_min = 1.332
    n2 = scaledata(PG2_gray, n_min, n_max)
    
    return n2
